File: java/Node.java
/* Pairing heap node */
import java.util.*;

public class Node<T extends Comparable> {
  Node<T> son;
  Node<T> sibling;
  T key;

  public Node(T key) {
    this.key = key;
  }

  public Node(T key, Node<T> son, Node<T> sibling) {
    this.key = key;
    this.son = son;
    this.sibling = sibling;
  }

  public Node<T> getSon() {
    return son;
  }

  public Node<T> getSibling() {
    return sibling;
  }

  public T getKey() {
    return key;
  }

  public Node<T> linkTo(Node<T> rhs) {
    if(rhs == null) {
      sibling = null;
      return this;
    }
    else {
      if(key.compareTo(rhs.key) > 0) {
        return rhs.linkTo(this);
      }

      rhs.sibling = son;
      son = rhs;
      sibling = null;

      return this;
    }
  }

  public Node<T> mergedSons() {
    Node<T> cur = son;

    if(cur == null)
      return null;
    else {
      Stack<Node<T>> nodes = new Stack<Node<T>>();

      while(cur != null && cur.sibling != null) {
        Node<T> next = cur.sibling.sibling;
        nodes.push(cur.linkTo(cur.sibling));
        cur = next;
      }

      if(cur != null)
        nodes.add(cur);

      while(nodes.size() > 1) {
        Node n1 = nodes.pop();
        Node n2 = nodes.pop();

        nodes.push(n2.linkTo(n1));
      }

      return nodes.pop();
    }
  }
}

File: csharp/Node.cs
/* Pairing heap node */

using System;
using System.Collections.Generic;

public class Node<T> where T : IComparable {
  public Node<T> Son { get; private set; }
  public Node<T> Sibling { get; private set; }
  public T Key { get; private set; }

  public Node(T key) {
    this.Key = key;
  }

  public Node<T> LinkTo( Node<T> rhs) {
    if(rhs == null) {
      Sibling = null;
      return this;
    }
    else {
      if(Key.CompareTo(rhs.Key) > 0) {
        return rhs.LinkTo(this);
      }

      rhs.Sibling = Son;
      Son = rhs;
      Sibling = null;

      return this;
    }
  }

  public Node<T> MergedSons() {
    Node<T> cur = Son;

    if(cur == null)
      return null;
    else {
      Stack<Node<T>> nodes = new Stack<Node<T>>();

      while(cur != null && cur.Sibling != null) {
        Node<T> next = cur.Sibling.Sibling;
        nodes.Push(cur.LinkTo(cur.Sibling));
        cur = next;
      }

      if(cur != null)
        nodes.Push(cur);

      while(nodes.Count > 1) {
        Node<T> n1 = nodes.Pop();
        Node<T> n2 = nodes.Pop();

        nodes.Push(n2.LinkTo(n1));
      }

      return nodes.Pop();
    }
  }
}
